### ISSUE TITLE
Optical Output DAC with realtek audio unlock the enhancements tab but no effects inside

### ISSUE THREAD

**Issue by ExoZNiK** [Falcosc/enable-loudness-equalisation#7]
I have a Dac connected by optical output to pc, with the guide it unlocks the enhancements tab (now it appear) in the sound panel of windows, but the are no enhancements to activate inside that. The devise use Realtek wich doesn't support the enhancements by default like nvidia high definition audio driver do. Is it normal that also with this script you can't use that with realtek device or there is to do something more?

**Comment by aarony1993** [comment]
Same here

https://i.imgur.com/ITMg64P.png

**Comment by Falcosc** [comment]
Interesting, it works on Realtek, Intel iGPU and Intel Arc.

Could you export HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\MMDevices\Audio\Render

Maybe the PreMixEffectClsid and PostMixEffectClsid is different for some driver versions.
Are you both on Windows 11?

I hope to find different IDs in your other sound devices because we use the following hardcoded ids:

https://github.com/Falcosc/enable-loudness-equalisation/blob/af818f9f2648c96b07d3c57b8669ed0d239ff70a/EnableLoudness.ps1#L52-L53

**Comment by ExoZNiK** [comment]
Yea i'm on windows 11 and i take a screenshot of the voices i think you were looking for:
![image](https://user-images.githubusercontent.com/73558501/234660713-990148ee-7af8-4c52-94ed-304216b0bece.png)

**Comment by ExoZNiK** [comment]
i don't have the 2 lines you mentioned so that's probably the problem and you figured out right :) any suggestion for a fix? or you have to implement the code now?

**Comment by Falcosc** [comment]
> i don't have the 2 lines you mentioned so that's probably the problem and you figured out right :) any suggestion for a fix? or you have to implement the code now?

We can look into the 36 audio devices you have installed on your computer. Maybe we find different MixEffectClsids there. But it is unlikely because the options appear even if these are not set. The only effect they have on my device is to actually apply the configured filters.

If you know the name of a device on your computer which has the features, we could try to figure out if there are some additional settings configured in there. But the likelihood that we find the missing piece is very low, since I already know that MixEffectClsids don't change what appears in the UI.

If the export of the whole folder doesn't give us any hints, I will add this "no features in layout" as a known issue for an unknown list of audio drivers.
Meanwhile, you could try out some different drivers, maybe there is one which brings these UI options where you only need the script to make the tab accessible.

**Comment by ExoZNiK** [comment]
How can i export that? i just use 2 device (monitor and optical output), so maybe are very old devices. i don't know how to differentiate them properly. i can give you a screenshot of the device where i want to make loudness equalization work and maybe obtain the id from that? 
![image](https://user-images.githubusercontent.com/73558501/234671945-b228c00d-4c49-4604-af59-c41dedbd98b3.png)
![image](https://user-images.githubusercontent.com/73558501/234672363-cd1818ff-7a78-4b25-8b77-a14f3a69c23b.png)
![image](https://user-images.githubusercontent.com/73558501/234673433-c99bd6d9-b3b4-443e-bc7e-0d242f91d323.png)
![image](https://user-images.githubusercontent.com/73558501/234673602-a3f775f1-6ca8-443b-823e-e10ad47dd6a0.png)
![image](https://user-images.githubusercontent.com/73558501/234673696-a372a52c-f293-403a-ab05-92764f8659a9.png)

tell me if i can help with something in particular

**Comment by Falcosc** [comment]
> How can i export that? 
Right-click on the registry folder and export, it will export all subfolders.

**Comment by ExoZNiK** [comment]
[realtek.zip](https://github.com/Falcosc/enable-loudness-equalisation/files/11336510/realtek.zip)
![image](https://user-images.githubusercontent.com/73558501/234679952-69074175-94ed-459e-981c-15e3ae2fe27f.png)
![image](https://user-images.githubusercontent.com/73558501/234680067-21611588-ddf5-46b0-9ab3-602d643d180d.png)
that's the zip with the reg file inside and the screen with the monitor diver (which has the loudness equalization avaible). ask if something is missing

**Comment by Falcosc** [comment]
2 ideas:
- we could try out the effects from your "Logitech G933 Gaming Headset" does it have any enhancements you need?
- PG279Q is an active audio output, did you try to enable the enhancements there? But it has no effects configured, so unlikely to work

The main problem is that you don't have any device with MixEffects. The trick works on drivers which provide effects for stuff like speakers, but don't expose the feature to other devices. In such a case the feature is there but just hidden.

But if there is no device using this feature, they may crafted a reduced feature driver package where effects are not included at all. In such case, you can't unhide them.

I don't know if "does only work if the driver does provide the functionality for some instead of all outputs" is the correct assumption. But I would like to add this somewhere in the documentation. What do you think, should this go to the known issues or maybe into the introduction?

**Comment by ExoZNiK** [comment]
PG279Q is the asus monitor and is with "nvidia high definiton audio driver" whihc has all the enchantements by default. i was trying to achieve loudness equalization on my "Realtek Digital Output" which is the one i sue (the dac with the headpones connected). g933 was my old gaming set i don't use that anymore. I was hoping that the ssumption you made was correct, if it was so the monitor has it and should unlock also realtek, but for my case seems that realtek is not compatible by default because it hasn't the function at all (not just hidden as you said). Pretty sad ;/

**Comment by ExoZNiK** [comment]
Maybe in the introduction readme file you can say that realtek is not compatible, as the other person have my problem too, and let this issue open (with some error label marked known) if you have any idea on how to fix that.

**Comment by Falcosc** [comment]
> Maybe in the introduction readme file you can say that realtek is not compatible

Unfortunately, that would be misleading because on some it does work. Maybe there are multiple versions of it? One with reduced features and one with full feature set?
For speaker outputs, you get these options by default, but only if you have a realtek driver package with enhancement support. 
So if you see enhancements for speakers many times, you will not see them for headphones, so this script is used to unhide the features available on speaker output for usage on headphones, digital outputs or monitor outputs.
Which means step 1 would be finding a driver which includes enhancements (if you don't have 3.5mm jack it is maybe a bit difficult to check.
But step 2 would be unhiding the options on any other device if you have them on the 3.5mm jack output

**Comment by ExoZNiK** [comment]
I tried also to connect the sennheiser hd599 (headphones) directly on the jack 3.5mm but it has the same problem (realtek driver also on that) the dac is connected by optical cable. I didn't find another driver compatibile with these device in te sound panel if i search on my computer, and if a reinstall from zero it intall the realtek. I never find online a person that have these enchantements on the realtek driver. That's why this script sound interesting :)

**Comment by Falcosc** [comment]
Windows ships a driver which is compatible with Realtek and has the features. But you are likely to lose the digital output there.
So yes, it is just a matter of which realtek compatible driver you install. If you find one which is compatible with your realtek configuration (you can have different outputs configured for the same realtek package number)

But even with the enhancement supporting driver, you will only be able to enable them for speakers. All other outputs need the unhide configuration.

**Comment by Falcosc** [comment]
You can just try any of the many guides which suggest replacing the realtek driver with the more generic Microsoft version of it by manually picking which driver you want to load in the update driver dialog.

**Comment by ExoZNiK** [comment]
![image](https://user-images.githubusercontent.com/73558501/234709934-4c5b9823-95bd-4a00-94e2-712754c84f19.png)
i have just this compatible by default and the usb 2.0 i think is the generic but never seemed good to me

**Comment by Simbiat** [comment]
I have similar situation, although I am using onboard audio with external speakers (with 3.5 jack). Digital audio from GPU has the enhancements visible, but the analog speakers - do not. MB is Asus TUF GAMING B650-PLUS and its audio drivers come with DTS, so maybe that is creating a conflict?
Attached are extracts from registry for both devices.

[Registry extracts.zip](https://github.com/Falcosc/enable-loudness-equalisation/files/11582533/Registry.extracts.zip)

**Comment by Falcosc** [comment]
No, conflicts are unlikely. It is more likely that your driver package has a reduced feature set.
You could try a different one directly from the audio chip producer instead of ASUS. But I can't tell you how to find out which driver packages contain the features you are looking. I only know how to make them visible after they are installed if they are there.

If you found one which ships loudness equalization for Audio Output X, I could try to find out why it isn't showing in your Audio Output Y.
If none of all your audio outputs have it, then it is unlikely to be part of the driver.

I copied the parameter from my 3.5 jack output to my HDMI output, but this only worked because it was there in the 3.5mm jack output. And that's what the script is doing

**Comment by Simbiat** [comment]
My HDMI (and DP) have loudness equalization, and 3.5 does not, that why I attached the registry exports/extracts for both outputs, since the above thread suggested, that you may know why it's not showing. Could you, please, take a look, when you have time?

**Comment by Falcosc** [comment]
That's sounds promising
How is the Name of the HDMI and DP Port Output and how is the 3.5mm output name?

All 3 using the same audio device?
Then it will be easy, likely just the wrong class id since this is one of the points I needed to guess and hard-code because I don't have a way of getting the correct Id.

Could you enable Loudness Equalization on your HDMI or DP device to have a complete reference in case the configuration isn't showing in the disabled state? If you already had it active for one of your monitors, activate it on the other to make it easier to spot the difference to make sure I got the right one, since I can't trial and error without your system on my hand.

**Comment by Simbiat** [comment]
HDMI had it enabled, when I exported the registry branch. In the .zip files I shared above amd.reg - is the branch for HDMI (and only the exact output I was using), and speakers.reg - is branch for 3.5 jack (and only it).

**Comment by Falcosc** [comment]
Did you enable it manually or with the script?

I need an HDMI export manually enabled and an HDMI export manually disabled

If you already executed the script, you can also use your DP device to make the on vs off difference export.

**Comment by Simbiat** [comment]
I enabled it manually for HDMI. Technically I did not run the script, I just used the registry values from it to manually add them.

**Comment by Falcosc** [comment]
@Simbiat found your issues. Your Displays are going over the AMD Driver and the Speaker over the Realtek driver. But the Realtek driver is known for not shipping loudness equalization. The generic microsoft driver with supports realtek and ships it.

But maybe we're addressing the realtek driver wrong.
You can take a look in your C:\Windows\INF\oem139.inf file (that's the current active copy of the string mappings) and check if your driver uses the same Ids for Key (PKEY_FX_) and value (FX_)
```
PKEY_FX_PreMixClsid  = "{D04E05A6-594B-4fb6-A80D-01AF5EED7D1D},1"
PKEY_FX_PostMixClsid = "{D04E05A6-594B-4fb6-A80D-01AF5EED7D1D},2"
PKEY_FX_UiClsid      = "{D04E05A6-594B-4fb6-A80D-01AF5EED7D1D},3"
FX_PREMIX_CLSID  = "{62dc1a93-ae24-464c-a43e-452f824c4250}"
FX_POSTMIX_CLSID = "{637c490d-eee3-4c0a-973f-371958802da2}"
FX_UI_CLSID      = "{5860E1C5-F95C-4a7a-8EC8-8AEF24F379A1}"
```

**Comment by Simbiat** [comment]
Interesting: I tried reinstalling the drivers from ASUS website (in case something went wrong during setup of the new system and they got replaced by something), and now I see 3 entries in registry, that are called "Speakers" and "Realtek", all 3 referencing different INF files. None of the INF files have FX keys in them though. Makes me wonder, that maybe... Some other renderer needs to be selected?... I've attached registry extracts for each renderer that references speakers and Realtek, their respective INF files and extract of all renderers (just in case).

I can use the "High Definition Audio Device" driver instead of Realtek/ASUS one, it has loudness normalization, as expected, but that breaks DTS which comes with the motherboard. While I can't say for sure if I even need it, I've built my new PC after my Lenovo Legion y740 17irhg burnt, and that one had Dolby Atmos for speakers which made it sound really good for a laptop (and had normalization in Dolby Access app, which worked perfectly). As such I want to give DTS a chance if possible.
Generic driver does have Virtual Surround enhancement and has Spatial Sound, too, but Dolby setup definitely doe snot sound as good (can't hear difference compared to Windows Sonic using Edifier R1700BT), but I can't even compare the sound properly, because I can't freely turn DTS on/off.

SO, yeah, this more like a "want", rather than "need", I am just curious if it can be done, and maybe during the process we can find something that will allow you to improve the script, who knows?

[RegistryAndINFs.zip](https://github.com/Falcosc/enable-loudness-equalisation/files/11617496/RegistryAndINFs.zip)
